cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java

```diff
@@ -96,14 +113,72 @@ public ArtifactClasses inspectArtifact(Id.Artifact artifactId, File artifactFile
+   *                                  actually an Application.
    */
   public ArtifactClasses inspectArtifact(Id.Artifact artifactId, File artifactFile,
-                                         ClassLoader parentClassLoader) throws IOException {
+                                         ClassLoader parentClassLoader) throws IOException, InvalidArtifactException {
+
+    List<ApplicationClass> appClasses = inspectApplications(artifactFile);
 
     try (PluginInstantiator pluginInstantiator = new PluginInstantiator(cConf, parentClassLoader)) {
       List<PluginClass> plugins = inspectPlugins(artifactId, artifactFile, pluginInstantiator);
-      return new ArtifactClasses(plugins);
+      return ArtifactClasses.builder().addApps(appClasses).addPlugins(plugins).build();
+    }
+  }
+
+  private List<ApplicationClass> inspectApplications(File artifactFile) throws IOException, InvalidArtifactException {
+    List<ApplicationClass> apps = Lists.newArrayList();
+
+    Location artifactLocation = Locations.toLocation(artifactFile);
+    Manifest manifest = BundleJarUtil.getManifest(artifactLocation);
+    if (manifest == null) {
+      return apps;
+    }
+    Attributes manifestAttributes = manifest.getMainAttributes();
+    if (manifestAttributes == null) {
+      return apps;
+    }
+
+    // right now we force users to include the application main class as an attribute in their manifest,
+    // which forces them to have a single application class.
+    // in the future, we may want to let users do this or maybe specify a list of classes or
+    // a package that will be searched for applications, to allow multiple applications in a single artifact.
+    String mainClassName = manifestAttributes.getValue(ManifestFields.MAIN_CLASS);
+    if (mainClassName != null) {
+      try (CloseableClassLoader artifactClassLoader =
+             artifactClassLoaderFactory.createClassLoader(Locations.toLocation(artifactFile))) {
+
+        Object appMain = artifactClassLoader.loadClass(mainClassName).newInstance();
+        if (!(appMain instanceof Application)) {
+          throw new InvalidArtifactException(String.format("Application main class is of invalid type: %s",
+            appMain.getClass().getName()));
+        }
+
+        Application app = (Application) appMain;
+
+        TypeToken typeToken = TypeToken.of(app.getClass());
+        TypeToken<?> resultToken = typeToken.resolveType(Application.class.getTypeParameters()[0]);
+        Type configType;
+        // if the user parameterized their template, like 'xyz extends ApplicationTemplate<T>',
+        // we can deserialize the config into that object. Otherwise it'll just be a Config
+        if (resultToken.getType() instanceof Class) {
+          configType = resultToken.getType();
+        } else {
+          configType = Config.class;
+        }
+        apps.add(new ApplicationClass(mainClassName, "", schemaGenerator.generate(configType)));
```

Comment:
this is so that people will know what to send for an app's config.  May not end up being exactly this, but wanted to at least put something here.